src/Container.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * Class Container
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay;
+
+use Psr\Container\ContainerInterface;
+use WCPay\Internal\DependencyManagement\DelegateContainer\WooContainer;
+use WCPay\Internal\DependencyManagement\ExtendedContainer;
+use WCPay\Internal\DependencyManagement\ServiceProvider\PaymentsServiceProvider;
+
+/**
+ * WCPay Dependency Injection Container.
+ *
+ * Wraps the ExtendedContainer implementation to only allow public access to
+ * certain methods of the internal container.
+ *
+ * The public methods of this class should be a subset of the ExtendedContainer
+ * methods, and follow their signatures. Avoid adding other methods here.
+ *
+ * During tests, the internal ExtendedContainer is available through
+ * the `wcpay_get_test_container()` method, allowing full manipulation.
+ */
+class Container implements ContainerInterface {
+	/**
+	 * Internal container instance.
+	 *
+	 * @var ExtendedContainer
+	 */
+	private $container;
+
+	/**
+	 * Initializes the container.
+	 *
+	 * Dependencies should not be provided during runtime, but will allow
+	 * mocking during tests. This is only needed for the container.
+	 *
+	 * @param WooContainer $woo_container The delegate container for WooCommerce (Optional).
+	 */
+	public function __construct( WooContainer $woo_container = null ) {
+		$this->container = new ExtendedContainer();
+
+		// Allow the container to be used as a dependency.
+		$this->container->addShared( static::class, $this );
+
+		// Add shared services.
+		$this->load_providers();
+
+		// Allow delegating unresolved queries to the WooCommerce container.
+		$this->container->delegate( $woo_container ?? new WooContainer() );
+	}
+
+	/**
+	 * Retrieves an instance of a given class.
+	 *
+	 * @param string $id The ID of the class to retrieve.
+	 * @return mixed
+	 */
```

Comment:
May use this dynamic return type? That can assist IDEs better. I got that idea from here https://phpstan.org/developing-extensions/dynamic-return-type-extensions and it seems to work well in PhpStorm when I check the change of `class-wc-payment-gateway-wcpay.php` in this PR. 

```suggestion
	 * @template T
	 * @param class-string<T> $id The ID of the class to retrieve.
	 * @return T
	 */
```

Reply:
It turns out this template annotation was also mentioned in Psalm https://psalm.dev/docs/annotating_code/templated_annotations/#param-class-stringt

And League Container itself has a commit to improve this one https://github.com/thephpleague/container/commit/388e992884ab75632d93dbaf4d29ec024d64b964

Reply:
Amazing suggestion, @htdat! Implemented in https://github.com/Automattic/woocommerce-payments/pull/6862/commits/0766124aecb621f6d781d49e9e3ab76415010dd1

By the way, I never knew this existed in PHP.